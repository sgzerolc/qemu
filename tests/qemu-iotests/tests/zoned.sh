#!/usr/bin/env bash
#
# Test zone management operations.
#

seq="$(basename $0)"
echo "QA output created by $seq"
status=1 # failure is the default!

_cleanup()
{
  _cleanup_test_img
  sudo rmmod null_blk
}
trap "_cleanup; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
. ./common.rc
. ./common.filter
. ./common.qemu

# This test only runs on Linux hosts with raw image files.
_supported_fmt raw
_supported_proto file
_supported_os Linux

QEMU_IO="build/qemu-io"
IMG="--image-opts driver=zoned_host_device,filename=/dev/nullb0"
QEMU_IO_OPTIONS=$QEMU_IO_OPTIONS_NO_FMT

echo "Testing a null_blk device"
echo "Simple cases: if the operations work"
sudo modprobe null_blk nr_devices=1 zoned=1
# hidden issues:
# 1. memory allocation error of "unaligned tcache chunk detected" when the nr_zone=1 in zone report
# 2. qemu-io: after report 10 zones, the program failed at double free error and exited.
echo "report the first zone"
sudo $QEMU_IO $IMG -c "zr 0 0 1"
echo "report: the first 10 zones"
sudo $QEMU_IO $IMG -c "zr 0 0 10"

echo "open the first zone"
sudo $QEMU_IO $IMG -c "zo 0 0x80000"
echo "report after:"
sudo $QEMU_IO $IMG -c "zr 0 0 1"
echo "open the last zone"
sudo $QEMU_IO $IMG -c "zo 0x3e70000000 0x80000"
echo "report after:"
sudo $QEMU_IO $IMG -c "zr 0x3e70000000 0 2"

echo "close the first zone"
sudo $QEMU_IO $IMG -c "zc 0 0x80000"
echo "report after:"
sudo $QEMU_IO $IMG -c "zr 0 0 1"
echo "close the last zone"
sudo $QEMU_IO $IMG -c "zc 0x3e70000000 0x80000"
echo "report after:"
sudo $QEMU_IO $IMG -c "zr 0x3e70000000 0 2"


echo "reset the second zone"
sudo $QEMU_IO $IMG -c "zrs 0x80000 0x80000"
echo "After resetting a zone:"
sudo $QEMU_IO $IMG -c "zr 0x80000 0 5"

# success, all done
echo "*** done"
rm -f $seq.full
status=0
