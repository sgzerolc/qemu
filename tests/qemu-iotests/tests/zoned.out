QA output created by zoned.sh
Testing a null_blk device:
Simple cases: if the operations work
(1) report the first zone:
start: 0x0, len 0x80000, cap 0x80000, wptr 0x0, zcond:1, [type: 2]

report the first 10 zones
start: 0x0, len 0x80000, cap 0x80000, wptr 0x0, zcond:1, [type: 2]
start: 0x80000, len 0x80000, cap 0x80000, wptr 0x80000, zcond:1, [type: 2]
start: 0x100000, len 0x80000, cap 0x80000, wptr 0x100000, zcond:1, [type: 2]
start: 0x180000, len 0x80000, cap 0x80000, wptr 0x180000, zcond:1, [type: 2]
start: 0x200000, len 0x80000, cap 0x80000, wptr 0x200000, zcond:1, [type: 2]
start: 0x280000, len 0x80000, cap 0x80000, wptr 0x280000, zcond:1, [type: 2]
start: 0x300000, len 0x80000, cap 0x80000, wptr 0x300000, zcond:1, [type: 2]
start: 0x380000, len 0x80000, cap 0x80000, wptr 0x380000, zcond:1, [type: 2]
start: 0x400000, len 0x80000, cap 0x80000, wptr 0x400000, zcond:1, [type: 2]
start: 0x480000, len 0x80000, cap 0x80000, wptr 0x480000, zcond:1, [type: 2]

report the last zone:
start: 0x1f380000, len 0x80000, cap 0x80000, wptr 0x1f380000, zcond:1, [type: 2]


(2) opening the first zone
report after:
start: 0x0, len 0x80000, cap 0x80000, wptr 0x0, zcond:3, [type: 2]

opening the second zone
report after:
start: 0x80000, len 0x80000, cap 0x80000, wptr 0x80000, zcond:3, [type: 2]

opening the last zone
report after:
start: 0x1f380000, len 0x80000, cap 0x80000, wptr 0x1f380000, zcond:3, [type: 2]


(3) closing the first zone
report after:
start: 0x0, len 0x80000, cap 0x80000, wptr 0x0, zcond:1, [type: 2]

closing the last zone
report after:
start: 0x1f380000, len 0x80000, cap 0x80000, wptr 0x1f380000, zcond:1, [type: 2]


(4) finishing the second zone
After finishing a zone:
start: 0x80000, len 0x80000, cap 0x80000, wptr 0x100000, zcond:14, [type: 2]


(5) resetting the second zone
After resetting a zone:
start: 0x80000, len 0x80000, cap 0x80000, wptr 0x80000, zcond:1, [type: 2]
*** done
